### ISSUE TITLE
thirdPartyCookiesEnabled on iOS

### ISSUE THREAD

**Issue by Martyneris** [react-native-community/react-native-webview#486]
Is there a way to enable third party cookies on iOS for Webview? 
My particular problem is that it is impossible to embed a twitter video inside a Webview unless all third party cookies have been accepted.

The prop thirdPartyCookiesEnabled is for android only and it states in documentation that iOS should enable the third party cookies by default.

Stack overflow doesn't provide much info on this as well, a couple of open tickets. 

Is there any workaround for this?

**Comment by persisgrio** [comment]
This is already enabled on iOS by default (according to docs [here](https://github.com/react-native-community/react-native-webview/blob/master/docs/Reference.md#thirdpartycookiesenabled)).  Do you have your `originWhiteList` set to accepting all (`[*]`)?

**Comment by Martyneris** [comment]
@persisgrio its supposed to be so, yes , I've read the documentation as well.
It did not work without your suggested originWhiteList prop and it still doesn't work after I added it and tested it out.

**Comment by persisgrio** [comment]
Are you seeing the twitter render at all? 

At minimum to have the tweet render at all (note: I tested on an iPhone X and iPhone 6), I needed to add some width to the parent container and a height to the WebView.  So the only props I absolutely needed on the WebView were `source`, `originWhiteList`, and `style`.  

Do you have a code snippet?

**Comment by Martyneris** [comment]
Yes, the render can be seen :  
- the Video player with a thumbnail 
- The play button, after pressing it I get the message about  twitter's use of cookies
- After pressing OK there's another Play button and after pressing it nothing happens, the player just freezes.
- the twitter post below the video player window

The snippet : 

`<WebView
          ref={(webview) => {
            this.webview = webview;
            playerRef(webview);
          }}
          onLoad={() => watch && this.playbackPlay()}
          source={{ uri: videoUrl }}
          scrollEnabled={false}
          injectedJavaScript={fixDefinedPostMessage + fixAllowedZoom}
          scalesPageToFit={false} // disable zooming on iOS
          pointerEvents="none" // disable zooming & text selection on Android
          style={[styles.playerContainer, !watch && styles.hidden]}
          startInLoadingState
          thirdPartyCookiesEnabled
          allowsInlineMediaPlayback
          mediaPlaybackRequiresUserAction={false}
          renderLoading={this.renderLoadingIndicator}
          renderError={this.renderError}
          userAgent={this.state.userAgent}
          originWhitelist={['*']}
          javaScriptEnabled
          javaScriptEnabledAndroid
          onNavigationStateChange={({ url, navigationType, loading }) => {
            // console.log({ url, navigationType, loading, ...others })
            const shouldOpenBrowser = resolvedUrl && resolvedUrl !== url;
            if (shouldOpenBrowser) {
              this.webview.stopLoading();
              Linking.openURL(url);
            }
            if (!resolvedUrl && !loading) {
              // resolvedUrl = url;
              this.setState({ resolvedUrl: url });
            }
          }}
        />`

**Comment by persisgrio** [comment]
Ah, I see that you are using uri/url, I thought you were using embedded HTML.  I have not had much luck using the uri, but these issues might be of use 
- https://github.com/react-native-community/react-native-webview/pull/181.  Probably something about headers.  

I am observing the same behavior you described, but on android.

**Comment by gorvinsky** [comment]
**I believe that thirdPartyCookiesEnabled is not enabled by default on iOS.**

I've been building a web app using React and while testing on Safari browser on Mac, I will face cross origin (CORS) issues when fetching an API from another domain say example.com. I can resolve this by unchecking "Prevent cross-site tracking" setting in Safari > Preferences... > Privacy. Unchecking it is essentially enabling third party cookies on Safari and that solved the CORS issues.

But when loaded from the webview on iOS, I will face the CORS issues again even though thirdPartyCookiesEnabled is supposed to be enabled by default on iOS. I tried setting it true in the webview component as well but it has no effect.

On Android, it's working fine and this is how I came to the conclusion that thirdPartyCookiesEnabled is not enabled by default on iOS unlike what was written in the documentation.

Do correct me if I'm wrong.

**Comment by gorvinsky** [comment]
So I was able to solve my own issue by modifying some of the files in **"node_modules/react-native-webview"**. Basically I created a new method called **"setCustomCookie"** to set the third party cookie or any cookie of your choice. Below are the changes made for **version 5.12.0** as well as a sample code to use the method at the bottom.

**index.d.ts**
```
   ...

   /**
    * Executes the JavaScript string.
    */
   injectJavaScript: (script: string) => void;

+  /**
+   * Set custom cookie. For iOS only.
+   */
+  setCustomCookie: (domain: string, name: string, value: string) => void;

   ...
```

**ios/RNCWKWebView.h**
```
   ...
   
   @property (nonatomic, assign) BOOL sharedCookiesEnabled;
+  @property (nonatomic, assign) NSString *customCookieDomain;
+  @property (nonatomic, assign) NSString *customCookieName;
+  @property (nonatomic, assign) NSString *customCookieValue;
   @property (nonatomic, assign) BOOL pagingEnabled;
   
   ...
   
   + (void)setClientAuthenticationCredential:(nullable NSURLCredential*)credential;
   - (void)postMessage:(NSString *)message;
   - (void)injectJavaScript:(NSString *)script;
   - (void)goForward;
   - (void)goBack;
   - (void)reload;
   - (void)stopLoading;
+  - (void)setCustomCookie:(NSDictionary*)properties;
   
   ...
```

**ios/RNCWKWebView.m**
```
   ...
   
   if (@available(iOS 11.0, *)) {
     // Set Cookies in iOS 11 and above, initialize websiteDataStore before setting cookies
     // See also https://forums.developer.apple.com/thread/97194
     // check if websiteDataStore has not been initialized before
     if(!_incognito && !_cacheEnabled) {
       wkWebViewConfig.websiteDataStore = [WKWebsiteDataStore nonPersistentDataStore];
     }
     for (NSHTTPCookie *cookie in [[NSHTTPCookieStorage sharedHTTPCookieStorage] cookies]) {
       [wkWebViewConfig.websiteDataStore.httpCookieStore setCookie:cookie completionHandler:nil];
     }
+    if (![_customCookieDomain isEqual: [NSNull null]] && ![_customCookieName isEqual: [NSNull null]] && ![_customCookieValue isEqual: [NSNull null]]) {
+      NSDictionary *properties = [NSDictionary dictionaryWithObjectsAndKeys:
+      _customCookieDomain, NSHTTPCookieDomain,
+      @"/", NSHTTPCookiePath,
+      _customCookieName, NSHTTPCookieName,
+      _customCookieValue, NSHTTPCookieValue,
+      nil];
+      NSHTTPCookie *customCookie = [NSHTTPCookie cookieWithProperties:properties];
+      [wkWebViewConfig.websiteDataStore.httpCookieStore setCookie:customCookie completionHandler:nil];
+    }
   } else {
   
   ...
   
   - (void)postMessage:(NSString *)message
   {
     NSDictionary *eventInitDict = @{@"data": message};
     NSString *source = [NSString
       stringWithFormat:@"window.dispatchEvent(new MessageEvent('message', %@));",
       RCTJSONStringify(eventInitDict, NULL)
     ];
     [self injectJavaScript: source];
   }
   
+  - (void)setCustomCookie:(NSDictionary*)properties
+  {
+    NSHTTPCookie *customCookie = [NSHTTPCookie cookieWithProperties:properties];
+    [self.webView.configuration.websiteDataStore.httpCookieStore setCookie:customCookie completionHandler:nil];
+  }
   
   ...
```

**ios/RNCWKWebViewManager.m**
```
   ...
   
   RCT_EXPORT_METHOD(stopLoading:(nonnull NSNumber *)reactTag)
   {
     [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, RNCWKWebView *> *viewRegistry) {
       RNCWKWebView *view = viewRegistry[reactTag];
       if (![view isKindOfClass:[RNCWKWebView class]]) {
         RCTLogError(@"Invalid view returned from registry, expecting RNCWKWebView, got: %@", view);
       } else {
         [view stopLoading];
       }
     }];
   }
   
+  RCT_EXPORT_METHOD(setCustomCookie:(nonnull NSNumber *)reactTag domain:(NSString *)domain name:(NSString *)name value:(NSString *)value)
+  {
+    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, RNCWKWebView *> *viewRegistry) {
+      RNCWKWebView *view = viewRegistry[reactTag];
+      if (![view isKindOfClass:[RNCWKWebView class]]) {
+        RCTLogError(@"Invalid view returned from registry, expecting RNCWKWebView, got: %@", view);
+      } else {
+        NSDictionary *properties = [NSDictionary dictionaryWithObjectsAndKeys:
+          domain, NSHTTPCookieDomain,
+          @"/", NSHTTPCookiePath,
+          name, NSHTTPCookieName,
+          value, NSHTTPCookieValue,
+          nil];
+        [view setCustomCookie:properties];
+      }
+    }];
+  }
   
   #pragma mark - Exported synchronous methods
   
   ...
```

**lib/WebView.ios.js**
```
   ...
   
   _this.injectJavaScript = function (data) {
     UIManager.dispatchViewManagerCommand(_this.getWebViewHandle(), _this.getCommands().injectJavaScript, [data]);
   };
+  /**
+   * Set custom cookie. For iOS only.
+   */
+  _this.setCustomCookie = function (domain, name, value) {
+    UIManager.dispatchViewManagerCommand(_this.getWebViewHandle(), _this.getCommands().setCustomCookie, [domain, name, value]);
+  }
   
   ...
```

**lib/WebView.ios.d.ts**
```
   ...
   
   injectJavaScript: (data: string) => void;
+  /**
+   * Set custom cookie. For iOS only.
+   */
+  setCustomCookie: (domain: string, name: string, value: string) => void;
   
   ...
```

**lib/WebViewTypes.d.ts**
```
   ...
   
   export interface WebViewCommands {
     goForward: Function;
     goBack: Function;
     reload: Function;
     stopLoading: Function;
     postMessage: Function;
     injectJavaScript: Function;
+    setCustomCookie: Function;
     loadUrl: Function;
   }
   
   ...
```

Once you have created the method, assuming that you have gotten the reference to the webview component say ``this.webview`` then you may use it like this:
```
const cookieDomain = 'http://example.com';
const cookieName = 'hello';
const cookieValue = 'world';
this.webview.setCustomCookie(cookieDomain, cookieName, cookieValue);
```

**Comment by DominicTobias** [comment]
I don't think they are enabled by default 🤔
![simulator_screenshot_E5D117A4-9926-4BBB-8093-73CD0CDD0EC4](https://user-images.githubusercontent.com/760314/206917883-c8ae043d-73da-42a6-b82e-cf47cbc9beb3.png)
